### PR TITLE
Less generic code for Vec allocations

### PR DESCRIPTION
**Pull request by kornelski**
Follow up to https://github.com/rust-lang/rust/pull/120504#issuecomment-1989826191 which hopefully makes compilation faster.

### CODE REVIEW

File: library/alloc/src/raw_vec.rs [hunk 1]
```diff
@@ -559,12 +576,10 @@ unsafe impl<#[may_dangle] T, A: Allocator> Drop for RawVec<T, A> {
 
 // Central function for reserve error handling.
 #[cfg(not(no_global_oom_handling))]
-#[inline]
-fn handle_reserve<T>(result: Result<T, TryReserveError>) -> T {
-    match result.map_err(|e| e.kind()) {
-        Ok(res) => res,
-        Err(CapacityOverflow) => capacity_overflow(),
-        Err(AllocError { layout, .. }) => handle_alloc_error(layout),
+fn handle_reserve_err(e: TryReserveError) -> ! {
```
Comment (joboet) [inline library/alloc/src/raw_vec.rs:579]: Tiny nit, just so that the function name rhymes with `TryReserveError` and `handle_alloc_error`
```suggestion
fn handle_reserve_error(e: TryReserveError) -> ! {
```

File: library/alloc/src/vec/mod.rs [hunk 2]
```diff
@@ -478,7 +478,7 @@ impl<T> Vec<T> {
     #[stable(feature = "rust1", since = "1.0.0")]
     #[must_use]
     pub fn with_capacity(capacity: usize) -> Self {
-        Self::with_capacity_in(capacity, Global)
+        Vec { buf: RawVec::with_capacity_in(capacity, Global), len: 0 }
```
Comment (joboet) [inline library/alloc/src/vec/mod.rs:481]: Why did you change this?

**Review by joboet — COMMENTED** [review]
Looks like there are some very small performance gains, and this makes sense to me from a readability standpoint.

Just two things, and then this should be good to go!

File: library/alloc/src/vec/mod.rs (hunk 2 above)
Comment (kornelski) [reply library/alloc/src/vec/mod.rs:481]: It removes another generic function to inline on a hot path. This is likely where majority of the improvement comes from.

Comment (joboet) [reply library/alloc/src/vec/mod.rs:481]: I'm not entirely convinced of this; the MIR inliner should be really good at these call-forwarding things. But I can experiment on that after this has merged.

Comment (joboet) [reply library/alloc/src/vec/mod.rs:481]: The MIR inliner should be really good at simple bodies like this, so I'm not convinced of that. Could you, for the sake of experimentation, undo this change?

**Review by joboet — CHANGES_REQUESTED** [review]

File: library/alloc/src/vec/mod.rs (hunk 2 above)
Comment (kornelski) [reply library/alloc/src/vec/mod.rs]: @joboet Are you satisfied with the result? I don't know if the small difference is a measurement noise or not.

**Comment by kornelski** [comment]
I've rebased and went back to the original version that had a better perf result.

**Comment by kornelski** [comment]
BTW, the previous PR #120504 has made it into the next beta, so this follow-up may need cherry-picking into the beta too.

**Comment by joboet** [comment]
Yeah, this seems fine to me, I can play around with it later.
@bors r+ rollup=never
@rustbot label +beta-nominated
(as it fixes a performance regression in beta)

File: library/alloc/src/raw_vec.rs [hunk 3]
```diff
@@ -559,12 +576,11 @@ unsafe impl<#[may_dangle] T, A: Allocator> Drop for RawVec<T, A> {
 
 // Central function for reserve error handling.
```
Comment (RalfJung) [inline library/alloc/src/raw_vec.rs:577]:
```suggestion
/// Central function for reserve error handling.
/// Deliberately non-generic to reduce codegen bloat!
```

File: library/alloc/src/raw_vec.rs [hunk 4]
```diff
@@ -114,7 +114,10 @@ impl<T> RawVec<T, Global> {
     #[must_use]
     #[inline]
     pub fn with_capacity(capacity: usize) -> Self {
-        handle_reserve(Self::try_allocate_in(capacity, AllocInit::Uninitialized, Global))
+        match Self::try_allocate_in(capacity, AllocInit::Uninitialized, Global) {
+            Ok(res) => res,
+            Err(err) => handle_reserve_error(err),
+        }
```
Comment (RalfJung) [inline library/alloc/src/raw_vec.rs:120]: I feel like this would be more concise and more readable with `unwrap_or_else`.

Comment (kornelski) [reply library/alloc/src/raw_vec.rs:120]: In this file there's a comment:

            // We avoid `unwrap_or_else` here because it bloats the amount of
            // LLVM IR generated.

Comment (RalfJung) [reply library/alloc/src/raw_vec.rs:120]: Ah, that wasn't visible in the diff, thanks.

**Comment by joboet** [comment]
r=me once CI comes in green.

**Comment by kornelski** [comment]
I've removed word "reserve" from helper function name to avoid confusing tests that look for absence of functions like `try_reserve`.

**Comment by joboet** [comment]
Cool! r=me once CI comes in green.

**Review by joboet — APPROVED** [review]
